Build failure with mtl-2.3

```
Building library for th-desugar-1.13..
[ 1 of 14] Compiling Language.Haskell.TH.Desugar.AST ( Language/Haskell/TH/Desugar/AST.hs, dist/build/Language/Haskell/TH/Desugar/AST.o, dist/build/Language/Haskell/TH/Desugar/AST.dyn_o )
[ 2 of 14] Compiling Language.Haskell.TH.Desugar.OMap ( Language/Haskell/TH/Desugar/OMap.hs, dist/build/Language/Haskell/TH/Desugar/OMap.o, dist/build/Language/Haskell/TH/Desugar/OMap.dyn_o )
[ 3 of 14] Compiling Language.Haskell.TH.Desugar.OMap.Strict ( Language/Haskell/TH/Desugar/OMap/Strict.hs, dist/build/Language/Haskell/TH/Desugar/OMap/Strict.o, dist/build/Language/Haskell/TH/Desugar/OMap/Strict.dyn_o )
[ 4 of 14] Compiling Language.Haskell.TH.Desugar.OSet ( Language/Haskell/TH/Desugar/OSet.hs, dist/build/Language/Haskell/TH/Desugar/OSet.o, dist/build/Language/Haskell/TH/Desugar/OSet.dyn_o )
[ 5 of 14] Compiling Language.Haskell.TH.Desugar.FV ( Language/Haskell/TH/Desugar/FV.hs, dist/build/Language/Haskell/TH/Desugar/FV.o, dist/build/Language/Haskell/TH/Desugar/FV.dyn_o )
[ 6 of 14] Compiling Language.Haskell.TH.Desugar.Util ( Language/Haskell/TH/Desugar/Util.hs, dist/build/Language/Haskell/TH/Desugar/Util.o, dist/build/Language/Haskell/TH/Desugar/Util.dyn_o )
[ 7 of 14] Compiling Language.Haskell.TH.Desugar.Subst ( Language/Haskell/TH/Desugar/Subst.hs, dist/build/Language/Haskell/TH/Desugar/Subst.o, dist/build/Language/Haskell/TH/Desugar/Subst.dyn_o )
[ 8 of 14] Compiling Language.Haskell.TH.Desugar.Reify ( Language/Haskell/TH/Desugar/Reify.hs, dist/build/Language/Haskell/TH/Desugar/Reify.o, dist/build/Language/Haskell/TH/Desugar/Reify.dyn_o )
[ 9 of 14] Compiling Language.Haskell.TH.Desugar.Core ( Language/Haskell/TH/Desugar/Core.hs, dist/build/Language/Haskell/TH/Desugar/Core.o, dist/build/Language/Haskell/TH/Desugar/Core.dyn_o )

Language/Haskell/TH/Desugar/Core.hs:1166:24: error:
    Variable not in scope: getAll :: a0 -> Bool
     |
1166 |       uni_pats <- fmap getAll $ concatMapM (fmap All . isUniversalPattern) pats'
     |                        ^^^^^^

Language/Haskell/TH/Desugar/Core.hs:1166:50: error:
    • Data constructor not in scope: All :: Bool -> a0
    • Perhaps you meant variable ‘all’ (imported from Data.Foldable)
     |
1166 |       uni_pats <- fmap getAll $ concatMapM (fmap All . isUniversalPattern) pats'
     |                                                  ^^^
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -15,6 +15,7 @@ Version 1.14 [????.??.??]
   runtime, just as all other forms of non-exhaustive expressions are desugared.
 * Fix a bug in which `expandType` would not expand closed type families when
   applied to arguments containing type variables.
+* Allow building with `mtl-2.3.*`.
 
 Version 1.13 [2021.10.30]
 -------------------------

--- a/Language/Haskell/TH/Desugar/Core.hs
+++ b/Language/Haskell/TH/Desugar/Core.hs
@@ -20,14 +20,16 @@ import Language.Haskell.TH.Syntax hiding (lift)
 
 import Control.Monad hiding (forM_, mapM)
 import qualified Control.Monad.Fail as Fail
+import Control.Monad.Trans (MonadTrans(..))
+import Control.Monad.Writer (MonadWriter(..), WriterT(..))
 import Control.Monad.Zip
-import Control.Monad.Writer hiding (forM_, mapM)
 import Data.Data (Data)
 import Data.Either (lefts)
 import Data.Foldable as F hiding (concat, notElem)
 import qualified Data.Map as M
 import Data.Map (Map)
 import Data.Maybe (isJust, mapMaybe)
+import Data.Monoid (All(..))
 import qualified Data.Set as S
 import Data.Set (Set)
 import Data.Traversable

--- a/th-desugar.cabal
+++ b/th-desugar.cabal
@@ -47,7 +47,7 @@ library
       ghc-prim,
       template-haskell >= 2.11 && < 2.19,
       containers >= 0.5,
-      mtl >= 2.1,
+      mtl >= 2.1 && < 2.4,
       ordered-containers >= 0.2.2,
       syb >= 0.4,
       th-abstraction >= 0.4 && < 0.5,

